markdown/postprocessors.py

```diff
@@ -73,37 +72,26 @@ class RawHtmlPostprocessor(Postprocessor):
 
     def run(self, text: str) -> str:
         """ Iterate over html stash and restore html. """
-        replacements = OrderedDict()
-        for i in range(self.md.htmlStash.html_counter):
-            html = self.stash_to_string(self.md.htmlStash.rawHtmlBlocks[i])
-            if self.isblocklevel(html):
-                replacements["<p>{}</p>".format(
-                    self.md.htmlStash.get_placeholder(i))] = html
-            replacements[self.md.htmlStash.get_placeholder(i)] = html
-
         def substitute_match(m: re.Match[str]) -> str:
-            key = m.group(0)
-
-            if key not in replacements:
-                if key[3:-4] in replacements:
-                    return f'<p>{ replacements[key[3:-4]] }</p>'
-                else:
-                    return key
-
-            return replacements[key]
-
-        if replacements:
+            if key := m.group(1):
+                wrapped = True
+            else:
+                key = m.group(2)
+                wrapped = False
+            if (key := int(key)) >= len(self.md.htmlStash.rawHtmlBlocks):
+                return m.group(0)
+            html = self.stash_to_string(self.md.htmlStash.rawHtmlBlocks[key])
+            if self.isblocklevel(html) or not wrapped:
+                return pattern.sub(substitute_match, html)
+            return pattern.sub(substitute_match, f"<p>{html}</p>")
+
+        if self.md.htmlStash.html_counter:
```

Comment:
Or we could *not* use `html_counter` and only rely on the actual list, `rawHtmlBlocks`:

```suggestion
        if self.md.htmlStash.rawHtmlBlocks:
```

Reply:
I was going to say let's be consistent and use the `html_counter` everywhere. But then I noticed that we don't actually document that attribute of the [class instance](https://python-markdown.github.io/reference/markdown/util/#markdown.util.HtmlStash). Although, there is no good reason to not have it documented. And in this case, I don't think there is any performance advantage to either check. I'm indifferent to this.

Reply:
Let's be consistent and use `html_counter`.